### PR TITLE
[KSM] support keep sampling mask

### PR DESCRIPTION
**Pull request by zeroRains**
## Motivation

本 PR 为 FastDeploy 实现 Keep Sampling Mask (KSM) 功能，用于在 top_p/top_k 采样过程中返回保留的词汇表索引列表（稀疏格式）。

当前推理引擎在执行 top_p/top_k 采样时，仅返回最终采样的 token ID，但不提供采样过程中的候选集合信息。这导致：
1. 可解释性不足：无法了解模型在每个 token 生成时考虑了哪些候选词
2. 调试困难：难以分析采样策略（如 top_p=0.9, top_k=50）的实际效果
3. 下游应用受限：无法基于候选集合实现自定义后处理逻辑logprobs 
4. 归一化不完整：返回的 logprobs 未基于截断后的候选集合重新归一化
 
本 PR 通过新增 sampling_mask 字段，记录每个 token 采样时保留的词汇表索引（稀疏格式），并提供基于候选集合的 logprobs 重归一化功能。

> :bulb: If this PR is a Cherry Pick, the PR title needs to follow the format by adding the [Cherry-Pick] label at the very beginning and appending the original PR ID at the end. For example, [Cherry-Pick][CI] Add check trigger and logic(#5191)

> :bulb: 如若此PR是Cherry Pick，PR标题需遵循格式，在最开始加上[Cherry-Pick]标签，以及最后面加上原PR ID，例如[Cherry-Pick][CI] Add check trigger and logic(#5191)

## Modifications

sampler.py 下新增_compute_sampling_mask方法
添加启动参数--enable-keep-sampling-mask
logprobs.py 下新增logz的renormalize函数，`logprobs_renormalize_with_logz`
pre_and_post_process.py的post_processs中调用renormalize函数

## Usage or Command

服务启动指令：
```
export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
MODEL_PATH="/root/paddlejob/tmpspace/GLM-4.5-Air/"
python -m fastdeploy.entrypoints.openai.api_server \
    --port 9293 \
    --host $(hostname -i) \
    --model "$MODEL_PATH" \
    --disable-custom-all-reduce \
    --tensor-parallel-size 8 \
    --max-model-len 131072 \
    --max-num-seqs 32 \
    --gpu-memory-utilization 0.9 \
    --graph-optimization-config '{"use_cudagraph":true}' \
    --enable-logprob \
    --enable-keep-sampling-mask \
    --speculative-config '{"method":"mtp","num_speculative_tokens":1,"num_model_steps":1,"model":"'$MODEL_PATH'"}'
```

## Accuracy Tests

yes

## Checklist

- [x] Add at least a tag in the PR title.
  - Tag list: [`[FDConfig]`,`[APIServer]`,`[Engine]`, `[Scheduler]`, `[PD Disaggregation]`, `[Executor]`, `[Graph Optimization]`, `[Speculative Decoding]`, `[RL]`, `[Models]`, `[Quantization]`, `[Loader]`, `[OP]`, `[KVCache]`, `[DataProcessor]`, `[BugFix]`, `[Docs]`, `[CI]`, `[Optimization]`, `[Feature]`, `[Benchmark]`, `[Others]`, `[XPU]`, `[HPU]`, `[GCU]`, `[DCU]`, `[Iluvatar]`, `[Metax]`]
  - You can add new tags based on the PR content, but the semantics must be clear.
- [x] Format your code, run `pre-commit` before commit.
- [x] Add unit tests. Please write the reason in this PR if no unit tests.
- [x] Provide accuracy results.
- [x] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

File: fastdeploy/model_executor/layers/sample/logprobs.py [hunk 1]
```diff
@@ -208,6 +206,10 @@ def build_output_logprobs(
         mask = idx < share_inputs["accept_num"].unsqueeze(1)
         token_ids = paddle.masked_select(share_inputs["accept_tokens"], mask)
 
+    # Adapt for sampling mask
+    if num_logprobs is None:
+        return None, None, output_logits
```
[the remark this answers is not in the document]
Comment (zeroRains) [reply fastdeploy/model_executor/layers/sample/logprobs.py:211]: 无影响，只有logprobs or sampling_mask为true的时候会执行这个函数，如果只有sampling_mask，仍然需要计算出logits，如果只有logrpobs，那也需要计算logits，我觉得不影响

File: fastdeploy/model_executor/layers/sample/sampler.py [hunk 2]
```diff
@@ -105,6 +106,153 @@ def padding_sampling_params(top_p, top_k, infer_seed, seq_lens_this_time, seq_le
...
+    # Z_K = sum(probs[i] * final_mask[i]) for each request i
+    # logZ_K = log(Z_K), with small constant to avoid log(0)
+    # ------------------------------------------------------------------
+    candidate_probs = paddle.where(final_mask, sorted_probs, paddle.zeros_like(sorted_probs))
+    z_k = candidate_probs.sum(axis=-1)  # [B]
+    logz_per_batch = paddle.log(z_k + 1e-10)  # [B], GPU
+
+    # Slice only the leading max_k columns on GPU — typically max_k << vocab_size.
+    # All outputs stay on GPU; D2H is deferred to save_output via async copy_.
+    indices_window = sorted_indices.slice([1], [0], max_k)  # [B, max_k]
```
[the remark this answers is not in the document]
Comment (zeroRains) [reply fastdeploy/model_executor/layers/sample/sampler.py:226]: item会显示触发同步，slice时使用tensor作为坐标应该不会触发

File: fastdeploy/output/token_processor.py [hunk 3]
```diff
@@ -734,6 +744,15 @@ def _process_batch_output(self):
             batch = self.output_tokens[1, 0]
             tokens = tokens[2 : batch + 2]
 
+        # Receive sampling constraints per request from ZMQ side-channel (if enabled).
+        # The worker sends a dict {batch_id: sparse_vocab_indices} each step,
+        # where the value is a list[int] or list[list[int]] of allowed token ids
+        sampling_masks_per_request = {}
+        if self.use_sampling_mask and not envs.FD_USE_GET_SAVE_OUTPUT_V1 and hasattr(self, "sampling_mask_zmq_server"):
```
[the remark this answers is not in the document]
Comment (zeroRains) [reply fastdeploy/output/token_processor.py:751]: token_processor在单独进程，可以一直等待

**Review by yuanlehome — APPROVED** [review]
